[Kaya3842] iP

Kaya3842 opened the pull request:
### **Liv the ta(x)sks collector**

> "The shorter way to do many things is to only do one thing at a time." - Mozart ([source](https://www.inc.com/dan-scalco/22-time-management-quotes-to-inspire-you-to-achieve-your-goals.html))

Always worry about having to many things to handle in life? Don't worry, Liv is here to alleviate your burden, by remembering your tasks for you. 👌

**Why** use Liv? It is:

- Command-based
- **Simple** command to learn and use
- _FAST_ and _SUPER EFFICIENT_
-  _Totally_ **free**

To start using Liv:
1. Download the file [here](https://github.com/nus-cs2103-AY2324S2/ip/pull/158)
2. Install the _app_
3. Run the _app_
4. **Enjoy!!**

Features:

- [X] Managing tasks

- [X] Managing deadlines

- [X] Managing events

- [ ] Reminder _(coming soon)_

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Kaya3842 committed:
Level-0

Kaya3842 committed:
Echo

Kaya3842 committed:
Add, List

Kaya3842 committed:
Add Task class

Kaya3842 committed:
Mark as Done

Kaya3842 committed:
Added type of tasks

Kaya3842 committed:
Automated Text UI Testing

Kaya3842 committed:
Handle Errors

Kaya3842 committed:
Add delete function + handle more errors

Kaya3842 committed:
Add Save function

Kaya3842 committed:
Merge branch 'branch-Level-7'

Kaya3842 committed:
BREAKING CHANGE: now store time as a LocalDateTime object instead of a String object

Kaya3842 committed:
Summary:
Style: Change the class structure to add more OOP
Description:
Extract new classes such as Parser and Storage to separate the logic. Transfer some functions from Liv class to new classes

Kaya3842 committed:
Add more classes to make more OOP

Completely separate the main class Liv into smaller classes.

Kaya3842 committed:
Liv.java: Remove old logic and add new logic

Rewrite the main class Liv.java using the new classes for more OOP.

Kaya3842 committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Liv.java: Remove old logic and add new logic
  Add more classes to make more OOP
  Summary: Style: Change the class structure to add more OOP Description: Extract new classes such as Parser and Storage to separate the logic. Transfer some functions from Liv class to new classes

Kaya3842 committed:
Organize into Packages

Divide the classes into suitable java packages.

Classes that are closely related are divided into the same package.

Kaya3842 committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Organize into Packages

Kaya3842 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

Kaya3842 committed:
Add build.gradle

Add Gradle to use on the later tasks of the project.

Kaya3842 committed:
Merge branch 'branch-A-Gradle'

* branch-A-Gradle:
  Add build.gradle

Kaya3842 committed:
Add Unit Test

Unit Test on two classes TaskList and Task to test the code.

Functions tested: addTask, markAsDone, getDisplayedString.

Kaya3842 committed:
Create a JAR File

Package the app as an executable JAR file for easier distribution.

Kaya3842 committed:
Add the find feature

The find feature is used to find task using a keyword.

This feature will help user manage tasks in the list easier.

Kaya3842 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add the find feature

Kaya3842 committed:
Add javadoc comments to some of the code

Add header comments to some classes.

Kaya3842 committed:
Add more Javadoc

Add header comments to most classes and methods

Kaya3842 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add more Javadoc
  Add javadoc comments to some of the code

# Conflicts:
#	data/liv.Liv.txt
#	src/main/java/liv/ui/Ui.java

Kaya3842 committed:
Fix the code to the coding standard

Change boolean changedData to hasChangedData.

Add more header comments.

Reorder some imports in classes to make them clearer.

Kaya3842 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fix the code to the coding standard

# Conflicts:
#	src/main/java/liv/processor/Command.java

Kaya3842 committed:
Fix javadoc

Fix some header comments to better describe the classes.

jyztintan commented on src/main/java/liv/Liv.java:
```diff
@@ -0,0 +1,55 @@
+        public void run() {
+        storage.loadDataFile();
+        ui.displayGreetCommand();
+        boolean isExit = false;
```
Great use of "is/has/was" for namign boolean variables!

jyztintan commented on src/main/java/liv/container/Storage.java:
```diff
@@ -0,0 +1,104 @@
+     * Reads and processes one line from the data file.
+     * @param line the line to be processed.
+     */
+    private void handleLineFromDataFile(String line) {
```
Good naming of methods using verbs in camelCase

jyztintan commented on src/main/java/liv/container/TaskList.java:
```diff
@@ -0,0 +1,69 @@
+/**
+ * A class that stores the tasks from users.
+ */
+public class TaskList {
```
I liked that you followed the naming convention of PascalCases for Java class names

jyztintan commented on src/main/java/liv/processor/MarkCommand.java:
```diff
@@ -0,0 +1,39 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui) throws LivException {
+        int trueIndex = index - 1;
+        Task task = TaskList.getTask(trueIndex);
+        boolean currentState = task.getStatus();
```
Intuitive naming of variables that follow camelCase and are easy to understand

jyztintan reviewed (commented):
Great job on your code quality naming, adhering to the convention styles!

KSSWSept20 commented on src/main/java/liv/Liv.java:
```diff
@@ -0,0 +1,55 @@
+ * The main class to run the chatbot.
+ */
+public class Liv {
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
```
I like how single-valued and multi-valued variables are distinguished clearly.

KSSWSept20 commented on src/main/java/liv/container/Storage.java:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * Saves the tasks currently on the list to the data file.
+     */
+    public void saveTaskToFile() {
```
I like the naming of the function which explains clearly to the reader about what it does.

KSSWSept20 commented on src/main/java/liv/task/Task.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
I like how the boolean variable is named using "is...".

KSSWSept20 commented on src/main/java/liv/processor/Parser.java:
```diff
@@ -0,0 +1,197 @@
+    public static Command parse(String input) throws LivException {
+        if (input.equalsIgnoreCase("bye")) {
+            return parseByeCommand();
+        } else if (input.equalsIgnoreCase("list")) {
+            return parseListCommand();
+        } else if (input.startsWith("mark")) {
+            return parseMarkOrUnmarkCommand(input, true);
+        } else if (input.startsWith("unmark")) {
+            return parseMarkOrUnmarkCommand(input, false);
+        } else if (input.startsWith("delete")) {
+            return parseDeleteCommand(input);
+        } else if (input.startsWith("find")) {
+            return parseFindCommand(input);
+        } else if (input.startsWith("todo")) {
+            return parseTodoCommand(input);
+        } else if (input.startsWith("deadline")) {
+            return parseDeadlineCommand(input);
+        } else if (input.startsWith("event")) {
+            return parseEventCommand(input);
```
I like how the clear naming of functions in the Parser class makes it easy to read when being called.

KSSWSept20 reviewed (commented):
Overall, LGTM, I found the naming of classes, variables and functions appropriate and easy to understand.
Great job! 🤩

Kaya3842 committed:
Add GUI features

Add a GUI to the chatbot using JavaFX technology.

Kaya3842 committed:
Update classes to match GUI logic

Update the logic in command and UI classes to use the GUI features

Kaya3842 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Update classes to match GUI logic
  Add GUI features

Kaya3842 committed:
Fix logic for GUI

The old logic is used for CLI and that logic design breaks in GUI.

To make the app work as intended in GUI, most of the main logic has to
be changed.

The Command.execute() method now takes in the Storage parameter. This
allows the app to store data when execute() is called instead of
run() in the CLI.

Since the GUI uses a different start() method from the CLI, this change
in design is to make the logic consistent for both interfaces.

Kaya3842 committed:
Use Assertions and fix exception message

The assert feature can document important assumptions at various
points in the code. Make the exception message clearer so the user can
understand what is wrong in the command.

The command that users input cannot be null. The number from the input
command is a positive integer. The time for an event cannot be null.
The keyword in the find command cannot be null.

The number in the input command must be a positive integer.
The exception message is changed to make that error clearer.
Add another case where the keyword from find is empty.

Kaya3842 committed:
Improve code quality

Improve code quality can help improve readability, making the logic
easier to follow and debug.

Use blank lines to separate different groups of logic in a method.

Change the method initialize() in the MainWindow class to follow SLAP.
Move the greetingMessage out of the method to make a single level
of abstraction.

Having a single level of abstraction in a method can improve
readability.

Kaya3842 committed:
Merge pull request #2 from Kaya3842/branch-A-Assertions

Use Assertions and fix exception message

Kaya3842 committed:
Merge branch 'master' into branch-A-CodeQuality

Kaya3842 committed:
Merge pull request #3 from Kaya3842/branch-A-CodeQuality

Improve code quality

Kaya3842 committed:
Add extension C-MassOps

Add the feature to perform tasks on multiple items in one command.

Change the methods in Parser class to take in more than one index for
one command.

Change the logic of some command classes to support mass operations.

Kaya3842 committed:
Add clear command and fix logic

Add a clear command to delete everything in the task list.

Fix the logic in Mark, Unmark and Delete command to support changing
multiple tasks in one command.

Kaya3842 committed:
Merge branch 'branch-C-MassOps'

Kaya3842 committed:
Fix Clear command

The clear command has to delete all tasks in the storage.

Add storage.saveTaskToFile() method in ClearCommand.execute() to
rewrite the task list to an empty task list.

Kaya3842 committed:
Add checkstyle and fix coding style

Add the config/checkstyle directory to set up checkstyle.

Change the build.gradle to inlcude checkstyle.

Fix the code to follow the coding style by fixing all errors from
checkstyle.

Kaya3842 committed:
Merge branch 'branch-A-Checkstyle'

Kaya3842 committed:
Improve GUI

Add background image and app icon image.

Customize the dialog box to have different color for the user and the
chatbot.

Add the style.css file to define the background image style and dialog
box style.

Kaya3842 committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Improve GUI

Kaya3842 committed:
Change Ui message

Let's add a more consistent personality to the chatbot.

Change some of the Ui messages to match the desired personality of the
chatbot.

Kaya3842 committed:
Merge branch 'branch-A-Personality'

* branch-A-Personality:
  Change Ui message

Kaya3842 committed:
Fix javadoc

Let's use AI tools to improve the javadoc documentation.

Add the missing javadoc in some methods using the template generated
by AI.

Change some of the methods javadoc for more detailed documentation
using AI.

Kaya3842 committed:
Add Ui.png

Add Ui.png as a representative screenshot of the product.

The Ui.png image is added to the docs folder.

Kaya3842 committed:
Fix build.gradle for jar release

Change the build.gradle file to make a fat jar for release.

Add some images for README.md.

Make the icon.png bigger.

Kaya3842 committed:
Update README.md

Add the user guide to the project. Change the old README.md from Duke
template to Liv user guide.
